webkit2/webkit2.uri-scheme-request.lisp

```diff
@@ -0,0 +1,95 @@
+;;; webkit2.uri-scheme-request.lisp --- bindings for WebKitURISchemeRequest
+
+;; This file is part of cl-webkit.
+;;
+;; cl-webkit is free software; you can redistribute it and/or modify
+;; it under the terms of the MIT license.
+;; See `COPYING' in the source distribution for details.
+
+;;; Code:
+
+(in-package #:webkit2)
+
+(defctype webkit-uri-scheme-request :pointer)
+
+(defcfun "webkit_uri_scheme_request_get_scheme" :string
+  (request webkit-uri-scheme-request))
+(export 'webkit-uri-scheme-request-get-scheme)
+
+(defcfun "webkit_uri_scheme_request_get_uri" :string
+  (request webkit-uri-scheme-request))
+(export 'webkit-uri-scheme-request-get-uri)
+
+(defcfun "webkit_uri_scheme_request_get_path" :string
+  (request webkit-uri-scheme-request))
+(export 'webkit-uri-scheme-request-get-path)
+
+(defcfun "webkit_uri_scheme_request_get_web-view" (g-object webkit-web-view)
+  (request webkit-uri-scheme-request))
+(export 'webkit-uri-scheme-request-get-web-view)
+
+(defcfun "webkit_uri_scheme_request_finish" :void
+  (request webkit-uri-scheme-request)
+  (stream :pointer)  ; XXX: GInputStream
+  (stream-length :int)
+  (contents :string))
+(export 'webkit-uri-scheme-request-finish)
+
+(defcfun ("webkit_uri_scheme_request_finish_error" %webkit-uri-scheme-request-finish-error) :void
+  (request webkit-uri-scheme-request)
+  (g-error :pointer))
+
+(defun webkit-uri-scheme-request-finish-error (request)
+  (glib:with-g-error (err)
+    (%webkit-uri-scheme-request-finish-error request err)))
+(export 'webkit-uri-scheme-request-finish-error)
+
+(cffi:defcallback uri-scheme-processed :void ((request :pointer) (user-data :pointer))
+  (let ((callback (find (cffi:pointer-address user-data) callbacks :key (function callback-id))))
+    (flet ((g-memory-input-stream-new-from-data (data-string)
+             (let* ((ffi-string (cffi:foreign-string-alloc data-string))
+                    ;; TODO: Can we rely on lisp's `length' here?
+                    ;; Will it give the same result as strlen on ffi-string?
```

Comment:
Unsure, but I think `length' understands unicode, so maybe you'd need to coerce this to a byte array first.

Reply:
Solution was quite simple: `cffi:foreign-string-alloc` returns the length of the resulting string as a second value :)